### PR TITLE
✨ NEW: Add simple typographic replacements

### PR DESCRIPTION
**Pull request by tsutsu3**
## Issue

#5 

## Summary

Implemented `replacements (Simple typographic replacements)` rule.

## Change

- Added a test to make sure that even a single capital letter is matched correctly.

### CODE REVIEW

File: markdown_it/rules_core/replacements.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from .state_core import StateCore
+from ..token import Token
+
+LOGGER = logging.getLogger(__name__)
+
+# TODO:
+# - fractionals 1/2, 1/4, 3/4 -> ½, ¼, ¾
+# - miltiplication 2 x 4 -> 2 × 4
+
+RARE_RE = r"\+-|\.\.|\?\?\?\?|!!!!|,,|--"
```
Comment (chrisjsewell) [inline markdown_it/rules_core/replacements.py:25]:
```suggestion
RARE_RE = re.compile(r"\+-|\.\.|\?\?\?\?|!!!!|,,|--")
```

It's a bit more performant to pre-compile, then use e.g. `RARE_RE.search`

File: markdown_it/rules_core/replacements.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+# Workaround for phantomjs - need regex without /g flag,
+# or root check will fail every second time
+# SCOPED_ABBR_TEST_RE = r"\((c|tm|r|p)\)"
+
+SCOPED_ABBR_RE = r"\((c|tm|r|p)\)"
```
Comment (chrisjsewell) [inline markdown_it/rules_core/replacements.py:31]:
```suggestion
SCOPED_ABBR_RE = re.compile(r"\((c|tm|r|p)\)", flags=re.IGNORECASE)
```

same as above

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
Awesome thanks! Just a minor nitpick

**Comment by chrisjsewell** [comment]
Also you can just `pip install pre-commit`, then `pre-commit run -all` to correctly format

File: markdown_it/rules_core/replacements.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+(r) (R) → ®
++- → ±
+(p) (P) -> §
+... → … (also ?.... → ?.., !.... → !..)
+???????? → ???, !!!!! → !!!, `,,` → `,`
+-- → &ndash;, --- → &mdash;
+"""
+import logging
+import re
+from typing import List
```
Comment (chrisjsewell) [inline markdown_it/rules_core/replacements.py:14]:
```suggestion
from typing import List, Match
```

File: markdown_it/rules_core/replacements.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+SCOPED_ABBR = {
+    "c": "©",
+    "r": "®",
+    "p": "§",
+    "tm": "™"
+}
+
+
+def replaceFn(match: re.Match):
```
Comment (chrisjsewell) [inline markdown_it/rules_core/replacements.py:41]:
```suggestion
def replaceFn(match: Match):
```

File: markdown_it/rules_core/replacements.py (hunk 3 above)
Comment (tsutsu3) [reply markdown_it/rules_core/replacements.py:14]: Thank you.

I was just in trouble.

**Comment by tsutsu3** [comment]
If I add this line to `conf.py`:

`nitpick_ignore = [("py:class", "Match")]`

sphinx warning goes away, is it ok to add it?

How can I get rid of the warning?

**Comment by chrisjsewell** [comment]
> sphinx warning goes away, is it ok to add it?

yep thats absolutely fine, I don't think there is another "fix"

**Comment by tsutsu3** [comment]
> > sphinx warning goes away, is it ok to add it?
> 
> yep thats absolutely fine, I don't think there is another "fix"

**Comment by tsutsu3** [comment]
I added `nitpick_ignore = [("py:class", "Match")]` in `conf.py`.

With python 3.7 there is no warning, but with python 3.6 the warning is still there.

```
/home/circleci/.local/lib/python3.6/site-packages/markdown_it/rules_core/replacements.py:docstring of markdown_it.rules_core.replacements.replaceFn:: WARNING: py:class reference target not found: Match[~ AnyStr]
generating indices...  genindex py-modindexdone
```

Hmm..., it's going to take some time to fix... 🤔 😕

**Comment by chrisjsewell** [comment]
looking good to me 👍 Is it ready for review?

**Comment by tsutsu3** [comment]
Should I put together a commit with a rebase?

**Review by chrisjsewell — APPROVED** [review]

**Comment by chrisjsewell** [comment]
yeh weird, it normally has the button here to update the branch

if you could rebase then thanks

**Comment by tsutsu3** [comment]
I checked "Allow edits by maintainers"

My pc env has changed. 😢 

I can't rebase, can you handle it for me? 🙏

**Comment by chrisjsewell** [comment]
sorted thanks @tsutsu3 😄

**Pull request merged**